### PR TITLE
Implement DNS utils for Windows

### PR DESCRIPTION
**Pull request by oskirby**
When relying on the Wireguard Go module to configure DNS, we find ourselves with a chicken-and-egg bug in Windows when multihop is enabled. The DNS server lists need to be provided when creating the interface, but that is typically done for the entry server, however, we aren't given the DNS configuration until we configure the exit server, which is configured last.

To resolve the issue, we implement the `DnsUtils` class for Windows and remove the DNS setup from the Go module, which conveniently takes us one step closer to completing #1526

Closes: #1673

### CODE REVIEW

**Comment by oskirby** [comment]
It seems that `SetInterfaceDnsSettings()` is only supported on sufficiently new builds of Windows 10, so this PR actually breaks the VPN for older versions of Windows. In fact, it doesn't even start the daemon correctly because the dynamic linker just explodes before running anything.

**Comment by oskirby** [comment]
For windows builds 1809 and earlier, it seems that the only way to configure DNS resolvers is to use `netsh`. So, if `GetProcAddress()` fails to lookup the native win32 API to configure DNS, we can provide a fallback that uses `netsh` in a `QProcess()` to do the same.

One possible piece that's missing in this approach is the configuration of the DNS search domains. In the Wireguard Go module, this is done by editing the registry, but it's not clear if we need to do this when we are only configuring the root domain.

This fallback path was tested in a VM running windows 10 build 1703.

File: src/platforms/windows/daemon/dnsutilswindows.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  if (ConvertInterfaceAliasToLuid((wchar_t*)ifname.utf16(), &luid) != 0) {
+    logger.error() << "Failed to resolve LUID for" << ifname;
+    return false;
+  }
+  m_luid = luid.Value;
+
+  logger.debug() << "Configuring DNS for" << ifname;
+  if (m_setInterfaceDnsSettingsProcAddr == nullptr) {
+    return updateResolversNetsh(resolvers);
+  } else {
```
Comment (bakulf) [inline src/platforms/windows/daemon/dnsutilswindows.cpp:52]: No `else` after a `return`.

File: src/platforms/windows/daemon/dnsutilswindows.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  return netsh.exitCode() == 0;
+}
+
+bool DnsUtilsWindows::restoreResolvers() {
+  if (m_luid == 0) {
+    return true;
+  }
+
+  QList<QHostAddress> empty;
+  if (m_setInterfaceDnsSettingsProcAddr != nullptr) {
```
Comment (bakulf) [inline src/platforms/windows/daemon/dnsutilswindows.cpp:174]: In the previous check, the order is the opposite. it was `if (a == nullptr)...` No needs to change.

File: src/platforms/windows/daemon/dnsutilswindows.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    return updateResolversWin32(empty);
+  } else {
```
Comment (bakulf) [inline src/platforms/windows/daemon/dnsutilswindows.cpp:176]: No `else` after a `return`

**Review by bakulf — APPROVED** [review]

File: src/platforms/windows/daemon/dnsutilswindows.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef DNSUTILSWINDOWS_H
+#define DNSUTILSWINDOWS_H
+
+#include "daemon/dnsutils.h"
+
+#include <QHostAddress>
+#include <QMap>
```
[the remark this answers is not in the document]
Comment (oskirby) [reply src/platforms/windows/daemon/dnsutilswindows.h:11]: That is a copy paste error that made it's way here from `dnsutilsmacos.h`. Sorry.

**Pull request merged**